File: packages/solver_core/src/solver_core/recipe/reduce.py
from solver_core.constants import intup, MAX_FLAVOR, MAX_TOTAL, EFFECTS


def apply_ingredient(flavors: intup, ingredient: str) -> list[int]:
    effect, candidates = EFFECTS[ingredient]
    new_flavors = [0] * 5
    for i in range(5):
        val = flavors[i] + effect[i]
        if val < 0:
            new_flavors[i] = 0
        elif val > MAX_FLAVOR:
            new_flavors[i] = MAX_FLAVOR
        else:
            new_flavors[i] = val
    total = sum(new_flavors)
    if total > MAX_TOTAL:
        idx = max(candidates, key=lambda i: new_flavors[i])
        new_flavors[idx] = max(0, new_flavors[idx] + MAX_TOTAL - total)
    return new_flavors


def flavors_match(f1: intup, f2: intup) -> bool:
    return all(f1[i] == f2[i] for i in range(5))


def swap_groups(
    grouped: list[tuple[str, int]], i: int, j: int
) -> list[tuple[str, int]]:
    grouped = grouped[:]  # shallow copy
    grouped[i], grouped[j] = grouped[j], grouped[i]
    return grouped


def is_valid_grouping(grouped: list[tuple[str, int]], target: intup) -> bool:
    current = [0] * 5
    for ing, cnt in grouped:
        for _ in range(cnt):
            if ing == "strong" and sum(current) > MAX_TOTAL - 10:
                return False
            current = apply_ingredient(tuple(current), ing)
    return flavors_match(tuple(current), target)


def reduce_recipe(recipe: list[str], target: intup) -> list[tuple[str, int]]:
    if not recipe:
        return []

    # Group consecutive identical ingredients
    grouped = []
    current, count = recipe[0], 1
    for i in range(1, len(recipe)):
        if recipe[i] == current:
            count += 1
        else:
            grouped.append((current, count))
            current, count = recipe[i], 1
    grouped.append((current, count))

    i = 0
    while i < len(grouped) - 2:
        curr = grouped[i][0]
        candidates = [(j, g) for j, g in enumerate(grouped) if g[0] == curr and j > i]
        for j, _ in candidates:
            new_grouped = swap_groups(grouped, i + 1, j)
            if is_valid_grouping(new_grouped, target):
                grouped = new_grouped
                i += 1
        i += 1

    # Merge adjacent identical ingredients
    merged = []
    prev_ing, acc_cnt = grouped[0]
    for ing, cnt in grouped[1:]:
        if ing == prev_ing:
            acc_cnt += cnt
        else:
            merged.append((prev_ing, acc_cnt))
            prev_ing, acc_cnt = ing, cnt
    merged.append((prev_ing, acc_cnt))

    return merged

File: web/frontend/src/solver_core/recipe/reduce.ts
import { EFFECTS, MAX_FLAVOR, MAX_TOTAL } from "../constants";

export function applyIngredient(flavors: number[], ingredient: string): number[] {
    const [effect, candidates] = EFFECTS[ingredient];
    const newFlavors = [0, 0, 0, 0, 0];
    for (let i = 0; i < 5; i++) {
        const val = flavors[i] + effect[i];
        if (val < 0) newFlavors[i] = 0;
        else if (val > MAX_FLAVOR) newFlavors[i] = MAX_FLAVOR;
        else newFlavors[i] = val;
    }
    const total = newFlavors.reduce((a, b) => a + b, 0);
    if (total > MAX_TOTAL) {
        const idx = candidates.reduce((best, i) => {
            return newFlavors[i] > newFlavors[best] ? i : best;
        }, candidates[0]);
        newFlavors[idx] = Math.max(0, newFlavors[idx] + MAX_TOTAL - total);
    }
    return newFlavors;
}

export function flavorsMatch(f1: number[], f2: number[]): boolean {
    return f1.every((val, i) => val === f2[i]);
}

export function swapGroups(
    grouped: Array<[string, number]>,
    i: number,
    j: number,
): Array<[string, number]> {
    const out = grouped.slice();
    const tmp = out[i];
    out[i] = out[j];
    out[j] = tmp;
    return out;
}

export function isValidGrouping(grouped: Array<[string, number]>, target: number[]): boolean {
    let current: number[] = [0, 0, 0, 0, 0];
    for (const [ing, cnt] of grouped) {
        for (let k = 0; k < cnt; k++) {
            const sum = current.reduce((a, b) => a + b, 0);
            if (ing === "strong" && sum > MAX_TOTAL - 10) return false;
            current = applyIngredient(current, ing);
        }
    }
    return flavorsMatch(current, target);
}

export function reduceRecipe(recipe: string[], target: number[]): Array<[string, number]> {
    if (recipe.length === 0) return [];

    let grouped = recipe.reduce<Array<[string, number]>>((acc, ingredient) => {
        const lastGroup = acc[acc.length - 1];
        if (lastGroup && lastGroup[0] === ingredient) {
            lastGroup[1]++;
        } else {
            acc.push([ingredient, 1]);
        }
        return acc;
    }, []);

    let cursor = 0;
    while (cursor < grouped.length - 2) {
        const curr = grouped[cursor][0];
        const candidateIndexes = grouped
            .map(([ingredient], idx) => (idx > cursor && ingredient === curr ? idx : -1))
            .filter((idx) => idx !== -1);

        for (const candidateIndex of candidateIndexes) {
            const newGrouped = swapGroups(grouped, cursor + 1, candidateIndex);
            if (isValidGrouping(newGrouped, target)) {
                grouped = newGrouped;
                cursor++;
            }
        }
        cursor++;
    }

    return grouped.reduce<Array<[string, number]>>((acc, [ingredient, count]) => {
        const lastGroup = acc[acc.length - 1];
        if (lastGroup && lastGroup[0] === ingredient) {
            lastGroup[1] += count;
        } else {
            acc.push([ingredient, count]);
        }
        return acc;
    }, []);
}
